src/model.js

```diff
@@ -3519,8 +3515,11 @@ Instead of specifying a Model, either:
       return result;
     }, {});
 
-    if (_.size(where) === 0) {
-      return this.constructor.options.whereCollection;
+    const primaryKeyAttribute = this.constructor.primaryKeyAttribute;
+    const isAutoGenerated = this.constructor.rawAttributes[primaryKeyAttribute]._autoGenerated;
+
+    if (primaryKeyAttribute && isAutoGenerated && this.isNewRecord === false) {
+      throw new Error('You attempted to query over a model that does not have a primary key explicitly defined this is not allowed');
```

Comment:
There is a misunderstanding regarding the "has no primary" part:

The goal is to ban using instance methods for models that have no primary at all. Not just that it hasn't been set on the model (i.e. not fetched), but that the table doesn't have a primary key column at all (`this.constructor.primaryKeyAttributes` is empty).

That being said you unearthed another case that needs to be handled too: If the user doesn't fetch the primary key when selecting an entity, then these methods cannot work either.

The fix can be done in two parts. First a check at the very start of this method that ensures a primary key is available:

```typescript
if (this.constructor.primaryKeyAttributes.length === 0) {
  throw new Error(`This model instance method needs to be able to identify the entity in a stable way, but the model does not have a primary key attribute definition. Either add a primary key to this model, or use one of the following alternatives:
  
- instance methods "save", "update", "decrement", "increment": Use the static "update" method instead.
- instance method "reload": Use the static "findOne" method instead.
- instance methods "destroy" and "restore": use the static "destroy" and "restore" methods instead.
  `.trim());
}
```

Then in `primaryKeyAttributes.reduce`: If any attribute is null or undefined, throw something along these lines:

```typescript
throw new Error(`This model instance method needs to be able to identify the entity in a stable way, but this model instance is missing the value of its primary key "${attribute}". Make sure that attribute was not excluded when retrieving the model from the database.`);
```